src/shared/sleep-config.c

```diff
@@ -135,6 +135,59 @@ int read_battery_capacity_percentage(void) {
         return battery_capacity;
 }
 
+static int get_battery_discharge_rate_filepath(char **ret_path) {
+        sd_id128_t machine_id;
+        int r;
+
+        assert(ret_path);
+
+        r = sd_id128_get_machine(&machine_id);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to get machine ID: %m");
+
+        *ret_path = path_join("/var/lib/systemd/sleep", SD_ID128_TO_STRING(machine_id), "battery_discharge_percentage_rate_per_hour");
```

Comment:
We should have the machine-id too I'm pretty sure - a different installation on the same hardware might have different settings that affect the drain. Hashing with the battery's identifiers sounds good.

Reply:
so, thinking more about this: i don't think the identity of the machine and battery should not be in the filename, but in the file contents (and use a fixed filename).

i.e. if we'd put it in the file name then we could support using the same image on 5 different systems and it would persist individually, but I doubt that's really something we should optimize for. I find it a lot more important that the number of state files we accumulate here is restricted. Thus, if people use the same image and boot it in 500 machines, one after the other, we should end up with one state file on disk, not 500 or more until the disk is full.

Hence, store the estimation in a file under a fixed filename, and then also store the battery and system identifiers inside the same file. When loading the file, verify that they match the local system, and if they don't match, ignore them.

That should be safe (as we won't use estimations from different systems on ours), and ensure we don't grow our state without limits, which is important I think.

Reply:
i need more clarity here on what is expected....
New to all this and the above comment is really intense for me to decode...

Reply:
concat(machine_id , product_serial) ?

Reply:
To sum it up : 
`manufacturer`, `model_name`, `serial_number`, `machine_id`, `product_id` -> concat and hash

In file : 
`hash` : `discharge_rate`

while reading:
generate system and battery identifiers hash again
Read the file and store the string.
Split string as `hash` and `dis_rate`
Match the hash from file with the one generated at the time of calling.
If they match, then check discharge rate is in sane range and then return `discharge_rate`

Reply:
`manufacturer`, `model_name`, `serial_number` might not be available in case of non-battery system like Desktop so we will keep that value `NULL` ?

`machine_id` and `product_id` are mandatory i guess because a system can not exist without that.

Reply:
yeah it's fine, the hash will match either way